### PR TITLE
enhance: load deltalogs on demand when doing compactions

### PR DESCRIPTION
**Pull request by tedxu**
See #37234

### CODE REVIEW

File: internal/datanode/compaction/merge_sort.go [hunk 1]
```diff
@@ -89,13 +78,41 @@ func mergeSortMultipleSegments(ctx context.Context,
...
+
+			ts, ok := segmentDelta[i][v.PK.GetValue()]
+			// insert task and delete task has the same ts when upsert
+			// here should be < instead of <=
+			// to avoid the upsert data to be deleted after compact
+			if ok && uint64(v.Timestamp) < ts {
+				continue
+			}
+			return v, nil
+		}
```
Comment (XuanYang-cn) [inline internal/datanode/compaction/merge_sort.go:108]: Please record `deletedRowCount` here

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

**Review by XuanYang-cn — APPROVED** [review]
/lgtm

**Review by congqixia — APPROVED** [review]
/approve

**Pull request merged**